main/src/library/LazyList.flix

```diff
@@ -123,4 +123,62 @@ namespace LazyList {
         case LCons(x, ys) => toListAcc(force ys, x :: acc)
         case LList(ys)    => toListAcc(force ys, acc)
     }
+
+    ///
+    /// Reverses the list `l`.
+    ///
+    pub def reverse(l: LazyList[a]): LazyList[a] = reverseAcc(ENil, l)
+
+    ///
+    /// Helper function for `reverse`.
+    ///
+    def reverseAcc(acc: LazyList[a], l: LazyList[a]): LazyList[a] = match l {
+        case ENil         => acc
+        case ECons(x, ys) => reverseAcc(ECons(x, acc), ys)
+        case LCons(x, ys) => reverseAcc(ECons(x, acc), force ys)
+        case LList(ys)    => reverseAcc(acc, force ys)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `l`.
+    ///
+    /// Whether `f` is applied eagerly or lazily depends on its purity:
+    ///
+    /// - If `f` is pure then it is applied lazily (i.e. the tail is not forced).
+    /// - If `f` is impure then it is applied eagerly (i.e. the entire list `l` is forced).
+    ///
+    pub def map(f: a -> b & ef, l: LazyList[a]): LazyList[b] & Impure =
+        matchEff f {
+            case Pure   => mapLazy(f as a -> b & Pure,   l)
+            case Impure => mapEager(f as a -> b & Impure, l)
+        }
+
+    ///
+    /// Returns the result of applying `f` to every element in `l`.
+    ///
+    /// Applies `f` lazily (i.e. the tail is not forced).
+    ///
+    def mapLazy(f: a -> b, l: LazyList[a]): LazyList[b] = match l {
+        case ENil         => ENil
+        case ECons(x, ys) => ECons(f(x), mapLazy(f, ys))
+        case LCons(x, ys) => LCons(f(x), lazy mapLazy(f, force ys))
+        case LList(ys)    => LList(lazy mapLazy(f, force ys))
+    }
```

Comment:
The usual naming is: ECons(x, xs) and similar, not `ys`. We use `y` and `ys` when we have two lists; then x and xs, and y and ys.